### DISCUSSION TITLE
Guidance for Uppy Companion (particularly for attaching by URL)

### DISCUSSION THREAD

**Discussion by Floppy** [shrinerb/shrine#693] [Q&A] [answered]
I'm using Uppy with Shrine, very happily. However, I'd really like to add the "URL" uploader that lets me upload just by pasting a link to the file. That uses Uppy Companion on the server side to do it.

There are a couple of mentions of Companion in the code and issues for Shrine, which imply that the presign endpoint can be used to support Companion requests somehow; but I'm not at all clear on how that works, or whether Shrine can even support what I want to do. 

Is there any guidance out there on how to properly get Shrine to pretend to be Companion? I really don't want to have to run another express app alongside my Ruby app...

**Comment by janko** [comment, marked as answer]
Shrine's presign endpoint and [uppy-s3_multipart](https://github.com/janko/uppy-s3_multipart) gem reimplement certain parts of Uppy companion in Ruby. The Ruby endpoints quack like the Uppy Companion endpoints, so Uppy doesn't mind.

It seems you would need to reimplement the `/url/meta` and `/url/get` endpoints from the Companion, here is the controller – https://github.com/transloadit/uppy/blob/main/packages/%40uppy/companion/src/server/controllers/url.js. Then you need to set `companionUrl` to your app in the `Url` plugin settings.

**Reply by Floppy** [reply]
Thanks, that's very useful - if I manage to get it working, I'll see if I can package it up as a gem like your multipart one. 👍🏻 

BTW, thanks for Shrine, it's been super helpful - I'm using it in https://manyfold.app :)
